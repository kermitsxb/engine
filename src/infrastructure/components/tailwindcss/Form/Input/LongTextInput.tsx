import { InputContext } from './BaseInput'
import type { LongTextInputProps } from '/domain/components/Form/Input/LongTextInput'

export const LongTextInput = ({
  field,
  label,
  description,
  placeholder,
  required,
}: LongTextInputProps) => {
  return (
    <InputContext label={label} description={description} field={field} required={required}>
      <textarea
        id={field}
        name={field}
        required={!!required}
        className="py-2 px-3 sm:py-3 sm:px-4 block w-full border-gray-200 rounded-lg sm:text-sm focus:border-blue-500 focus:ring-blue-500 disabled:opacity-50 disabled:pointer-events-none dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:placeholder-neutral-500 dark:focus:ring-neutral-600"
        rows={3}
        placeholder={placeholder}
        data-hs-textarea-auto-height=""
      ></textarea>
    </InputContext>
  )
}
