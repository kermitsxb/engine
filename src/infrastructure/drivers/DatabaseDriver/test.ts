import type { IDatabaseTableDriver } from '@adapter/spi/drivers/DatabaseTableSpi'
import BunTest, { afterAll } from 'bun:test'

export function testDatabaseTableDriver(
  { describe, beforeAll, it, expect }: typeof BunTest,
  setup: (date: Date) => Promise<IDatabaseTableDriver>,
  teardown?: () => Promise<void>
) {
  let databaseTable: IDatabaseTableDriver
  const date = new Date()

  beforeAll(async () => {
    databaseTable = await setup(date)
  })

  afterAll(async () => {
    if (teardown) await teardown()
  })

  describe('read', () => {
    it('should return a row with a filter', async () => {
      // WHEN
      const row = await databaseTable.read({
        field: 'name',
        operator: 'Is',
        value: 'John Doe',
      })

      // THEN
      expect(row).toStrictEqual({
        id: '1',
        fields: { name: 'John Doe' },
        created_at: date,
        updated_at: date,
      })
    })

    it('should not return a row with an empty "or" filter', async () => {
      // WHEN
      const row = await databaseTable.read({ or: [] })

      // THEN
      expect(row).toBeUndefined()
    })

    it('should not return a row with an empty "and" filter', async () => {
      // WHEN
      const row = await databaseTable.read({ and: [] })

      // THEN
      expect(row).toBeUndefined()
    })

    it('should not return a row with a filter', async () => {
      // WHEN
      const row = await databaseTable.read({
        field: 'name',
        operator: 'Is',
        value: 'Jane Doe',
      })

      // THEN
      expect(row).toBeUndefined()
    })

    it('should not return a row', async () => {
      // WHEN
      const row = await databaseTable.read({
        field: 'name',
        operator: 'Is',
        value: 'Jane Doe',
      })

      // THEN
      expect(row).toBeUndefined()
    })
  })

  describe('list', () => {
    it('should return a list of rows with no filter', async () => {
      // WHEN
      const rows = await databaseTable.list()

      // THEN
      expect(rows).toStrictEqual([
        {
          id: '1',
          fields: { name: 'John Doe' },
          created_at: date,
          updated_at: date,
        },
      ])
    })

    it('should return a list of rows with an empty "or" filter', async () => {
      // WHEN
      const rows = await databaseTable.list({ or: [] })

      // THEN
      expect(rows).toHaveLength(1)
    })

    it('should return a list of rows with an empty "and" filter', async () => {
      // WHEN
      const rows = await databaseTable.list({ and: [] })

      // THEN
      expect(rows).toHaveLength(1)
    })

    it('should return an empty list of rows with a filter', async () => {
      // WHEN
      const rows = await databaseTable.list({
        field: 'name',
        operator: 'Is',
        value: 'Jane Doe',
      })

      // THEN
      expect(rows).toHaveLength(0)
    })
  })
}
