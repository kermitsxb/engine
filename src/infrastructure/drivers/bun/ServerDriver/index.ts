import type { IServerDriver } from '/adapter/spi/drivers/ServerSpi'
import type { DeleteDto, GetDto, PatchDto, PostDto, RequestDto } from '/adapter/spi/dtos/RequestDto'
import type { ServerConfig } from '/domain/services/Server'
import type { Response } from '/domain/entities/Response'
import { ElysiaDriver } from './ElysiaDriver'

export class ServerDriver implements IServerDriver {
  private _server: ElysiaDriver

  constructor(config: ServerConfig) {
    this._server = new ElysiaDriver(config)
  }

  get = async (path: string, handler: (getDto: GetDto) => Promise<Response>) => {
    await this._server.get(path, handler)
  }

  post = async (path: string, handler: (postDto: PostDto) => Promise<Response>) => {
    await this._server.post(path, handler)
  }

  patch = async (path: string, handler: (patchDto: PatchDto) => Promise<Response>) => {
    await this._server.patch(path, handler)
  }

  delete = async (path: string, handler: (deleteDto: DeleteDto) => Promise<Response>) => {
    await this._server.delete(path, handler)
  }

  notFound = async (handler: (requestDto: RequestDto) => Promise<Response>) => {
    await this._server.notFound(handler)
  }

  start = async (): Promise<number> => {
    return this._server.start()
  }

  stop = async () => {
    await this._server.stop()
  }
}
