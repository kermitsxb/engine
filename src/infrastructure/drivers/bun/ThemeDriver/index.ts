import type { IThemeDriver } from '/adapter/spi/drivers/ThemeSpi'
import type { ThemeConfig } from '/domain/services/Theme'
import { TailwindCSSDriver } from './TailwindCSSDriver'

export class ThemeDriver implements IThemeDriver {
  private _driver: IThemeDriver

  constructor(config: ThemeConfig) {
    const { type } = config
    switch (type) {
      case 'tailwindcss':
        this._driver = new TailwindCSSDriver(config)
        break
      default:
        throw new Error(`Unsupported Theme driver: ${type}`)
    }
  }

  buildCss = async (): Promise<string> => {
    return this._driver.buildCss()
  }

  buildJs = async (): Promise<string> => {
    return this._driver.buildJs()
  }
}
