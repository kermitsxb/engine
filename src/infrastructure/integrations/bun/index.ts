import type { Integrations } from '@adapter/spi/integrations'
import type { NotionConfig } from '@domain/integrations/Notion'
import type { PappersConfig } from '@domain/integrations/Pappers'
import type { QontoConfig } from '@domain/integrations/Qonto'
import type { NgrokConfig } from '@domain/integrations/Ngrok'

import { NotionIntegration } from '../NotionIntegration'
import { PappersBunIntegration } from './PappersBunIntegration'
import { QontoBunIntegration } from './QontoBunIntegration'
import { NgrokBunIntegration } from './NgrokBunIntegration'

export const integrations: Integrations = {
  notion: (config?: NotionConfig) => new NotionIntegration(config),
  pappers: (config?: PappersConfig) => new PappersBunIntegration(config),
  qonto: (config?: QontoConfig) => new QontoBunIntegration(config),
  ngrok: (config?: NgrokConfig) => new NgrokBunIntegration(config),
}
