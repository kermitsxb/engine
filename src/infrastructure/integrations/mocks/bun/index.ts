import type { Integrations } from '@adapter/spi/integrations'
import type { NotionConfig } from '@domain/integrations/Notion'
import type { PappersConfig } from '@domain/integrations/Pappers'
import type { QontoConfig } from '@domain/integrations/Qonto'
import type { NgrokConfig } from '@domain/integrations/Ngrok'

import { NotionBunIntegration } from './NotionIntegration'
import { PappersBunIntegration } from './PappersIntegration'
import { QontoBunIntegration } from './QontoIntegration'
import { NgrokBunIntegration } from './NgrokIntegration'

export const mocks: Integrations = {
  notion: (config?: NotionConfig) => new NotionBunIntegration(config),
  pappers: (config?: PappersConfig) => new PappersBunIntegration(config),
  qonto: (config?: QontoConfig) => new QontoBunIntegration(config),
  ngrok: (config?: NgrokConfig) => new NgrokBunIntegration(config),
}
