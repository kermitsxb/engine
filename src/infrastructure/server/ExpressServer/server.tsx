import React from 'react'
import { IServerAdapter, ApiRoute, PageRoute } from '@adapter/spi/server/IServerAdapter'
import express, { Express } from 'express'
import { Server } from 'http'
import ReactDOMServer from 'react-dom/server'
import path from 'path'
import fs from 'fs-extra'
import { App } from '@domain/entities/app/App'
import { RequestDto, RequestQueryDto } from '@adapter/spi/server/dtos/RequestDto'
import { PageMapper } from '@adapter/api/page/mappers/PageMapper'
import { TableMapper } from '@adapter/api/table/mappers/TableMapper'
import { TableDto } from '@adapter/api/table/dtos/TableDto'
import { PageDto } from '@adapter/api/page/dtos/PageDto'

export interface FoundationData {
  page: PageDto
  tables: TableDto[]
  params: { [key: string]: string }
  adapters: {
    uiName: string
  }
}

export class ExpressServer implements IServerAdapter {
  private express?: Express
  private app?: App
  private server?: Server

  constructor(
    private _port: number,
    private _uiName: string
  ) {}

  initConfig(app: App) {
    this.express = express()
    this.express.use(express.json())
    this.express.use(express.urlencoded({ extended: true }))
    this.express.use(express.static('dist/public'))
    this.app = app
  }

  get port(): number {
    return this._port
  }

  async configureTables(routes: ApiRoute[]) {
    if (!this.express) throw new Error('Express not initialized')
    for (const route of routes) {
      switch (route.method) {
        case 'GET':
          this.express.get(route.path, async (req, res) => {
            const options: RequestDto = {
              method: req.method,
              path: req.url.split('?')[0],
              params: req.params,
            }
            if (req.query) {
              options.query = Object.keys(req.query).reduce((acc: RequestQueryDto, key: string) => {
                acc[key] = String(req.query[key])
                return acc
              }, {})
            }
            const { status = 200, json } = await route.handler(options)
            res.status(status).json(json)
          })
          break
        case 'POST':
          this.express.post(route.path, async (req, res) => {
            const { status = 200, json } = await route.handler({
              method: req.method,
              path: req.url,
              params: req.params,
              body: req.body,
            })
            res.status(status).json(json)
          })
          break
        case 'PATCH':
          this.express.patch(route.path, async (req, res) => {
            const { status = 200, json } = await route.handler({
              method: req.method,
              path: req.url,
              params: req.params,
              body: req.body,
            })
            res.status(status).json(json)
          })
          break
        case 'DELETE':
          this.express.delete(route.path, async (req, res) => {
            const { status = 200, json } = await route.handler({
              method: req.method,
              path: req.url,
              params: req.params,
            })
            res.status(status).json(json)
          })
          break
        default:
          break
      }
    }
  }

  async configurePages(routes: PageRoute[]) {
    if (!this.express) throw new Error('Express not initialized')
    this.express.get('/bundle.js', async (req, res) => {
      const bundle = await fs.readFile(path.resolve(__dirname, '../../../../bundle.js'), 'utf8')
      return res.send(bundle)
    })
    for (const route of routes) {
      this.express.get(route.path, async (req, res) => {
        const Page = await route.handler({ path: req.url.split('?')[0], params: req.params })
        const pageHtml = ReactDOMServer.renderToString(<Page />)
        if (!this.app) throw new Error('App not initialized')
        const page = this.app.getPageByPath(route.path)
        const data: FoundationData = {
          page: PageMapper.toDto(page),
          params: req.params,
          tables: TableMapper.toDtos(this.app.tables),
          adapters: { uiName: this._uiName },
        }
        const html = `
          <!DOCTYPE html>
          <html>
            <head>
              <title>${route.title}</title>
              <script>
                window.__FOUNDATION_DATA__ = ${JSON.stringify(data)}
              </script>
              <link href="/output.css" rel="stylesheet">
            </head>
            <body>
              <div id="root">${pageHtml}</div>
              <script src="/bundle.js"></script>
            </body>
          </html>
        `
        res.send(html)
      })
    }
  }

  async configureStorage(routes: ApiRoute[]) {
    if (!this.express) throw new Error('Express not initialized')
    for (const route of routes) {
      switch (route.method) {
        case 'GET':
          this.express.get(route.path, async (req, res) => {
            const options: RequestDto = {
              method: req.method,
              path: req.url.split('?')[0],
              params: req.params,
            }
            const { status = 200, file, headers = {}, json = {} } = await route.handler(options)
            res.status(status)
            if (status === 200 && file) {
              Object.keys(headers).forEach((key) => {
                res.setHeader(key, headers[key])
              })
              res.sendFile(file, { headers })
            } else {
              res.json(json)
            }
          })
          break
        default:
          break
      }
    }
  }

  async start(): Promise<void> {
    await new Promise((resolve, rejects) => {
      if (!this.express) throw new Error('Express not initialized')
      this.server = this.express.listen(this.port, () => {
        resolve(true)
      })
      this.server.on('error', async (error: unknown) => {
        rejects({ error })
      })
    })
  }

  async stop(): Promise<void> {
    await new Promise((resolve) => {
      if (!this.server) throw new Error('Server not initialized')
      this.server.close(() => {
        resolve(true)
      })
    })
  }
}
