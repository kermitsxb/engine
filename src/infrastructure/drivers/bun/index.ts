import type { Drivers } from '/adapter/spi/drivers'
import type { DatabaseConfig } from '/domain/services/Database'
import type { MonitorsConfig } from '/domain/services/Monitor'
import type { ServerConfig } from '/domain/services/Server'
import type { ThemeConfig } from '/domain/services/Theme'

import { DatabaseDriver } from './DatabaseDriver'
import { MonitorDriver } from './MonitorDriver'
import { ServerDriver } from './ServerDriver'
import { ThemeDriver } from './ThemeDriver'

import { drivers as common } from '../common'

export const drivers: Drivers = {
  ...common,
  monitor: (config: MonitorsConfig) => new MonitorDriver(config),
  database: (config: DatabaseConfig) => new DatabaseDriver(config),
  server: (config: ServerConfig) => new ServerDriver(config),
  theme: (config: ThemeConfig) => new ThemeDriver(config),
}
