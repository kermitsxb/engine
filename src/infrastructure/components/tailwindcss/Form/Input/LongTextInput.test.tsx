import { it, expect } from 'bun:test'
import { LongTextInput } from './LongTextInput'
import { renderToString } from 'react-dom/server'

it('should match the textarea snapshot', async () => {
  // GIVEN
  const input = <LongTextInput field="name" />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the textarea snapshot with label', async () => {
  // GIVEN
  const input = <LongTextInput field="name" label="Name" />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the textarea snapshot with description', async () => {
  // GIVEN
  const input = <LongTextInput field="name" description="Description" />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the textarea snapshot with label and description', async () => {
  // GIVEN
  const input = <LongTextInput field="name" label="Name" description="Description" />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the textarea snapshot with placeholder', async () => {
  // GIVEN
  const input = <LongTextInput field="name" placeholder="Placeholder" />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the textarea snapshot with required', async () => {
  // GIVEN
  const input = <LongTextInput field="name" required />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})
