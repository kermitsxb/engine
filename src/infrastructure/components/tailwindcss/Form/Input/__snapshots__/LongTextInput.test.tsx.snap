// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the textarea snapshot 1`] = `"<div><textarea id="name" name="name" class="py-2 px-3 sm:py-3 sm:px-4 block w-full border-gray-200 rounded-lg sm:text-sm focus:border-blue-500 focus:ring-blue-500 disabled:opacity-50 disabled:pointer-events-none dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:placeholder-neutral-500 dark:focus:ring-neutral-600" rows="3" data-hs-textarea-auto-height=""></textarea></div>"`;

exports[`should match the textarea snapshot with label 1`] = `"<div><label for="name" class="block mb-2 text-sm text-gray-700 font-medium dark:text-white">Name</label><textarea id="name" name="name" class="py-2 px-3 sm:py-3 sm:px-4 block w-full border-gray-200 rounded-lg sm:text-sm focus:border-blue-500 focus:ring-blue-500 disabled:opacity-50 disabled:pointer-events-none dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:placeholder-neutral-500 dark:focus:ring-neutral-600" rows="3" data-hs-textarea-auto-height=""></textarea></div>"`;

exports[`should match the textarea snapshot with description 1`] = `"<div><p class="mb-2 text-sm text-gray-500 dark:text-neutral-500">Description</p><textarea id="name" name="name" class="py-2 px-3 sm:py-3 sm:px-4 block w-full border-gray-200 rounded-lg sm:text-sm focus:border-blue-500 focus:ring-blue-500 disabled:opacity-50 disabled:pointer-events-none dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:placeholder-neutral-500 dark:focus:ring-neutral-600" rows="3" data-hs-textarea-auto-height=""></textarea></div>"`;

exports[`should match the textarea snapshot with label and description 1`] = `"<div><label for="name" class="block mb-2 text-sm text-gray-700 font-medium dark:text-white">Name</label><p class="mb-2 text-sm text-gray-500 dark:text-neutral-500">Description</p><textarea id="name" name="name" class="py-2 px-3 sm:py-3 sm:px-4 block w-full border-gray-200 rounded-lg sm:text-sm focus:border-blue-500 focus:ring-blue-500 disabled:opacity-50 disabled:pointer-events-none dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:placeholder-neutral-500 dark:focus:ring-neutral-600" rows="3" data-hs-textarea-auto-height=""></textarea></div>"`;

exports[`should match the textarea snapshot with placeholder 1`] = `"<div><textarea id="name" name="name" class="py-2 px-3 sm:py-3 sm:px-4 block w-full border-gray-200 rounded-lg sm:text-sm focus:border-blue-500 focus:ring-blue-500 disabled:opacity-50 disabled:pointer-events-none dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:placeholder-neutral-500 dark:focus:ring-neutral-600" rows="3" placeholder="Placeholder" data-hs-textarea-auto-height=""></textarea></div>"`;

exports[`should match the textarea snapshot with required 1`] = `"<div><textarea id="name" name="name" required="" class="py-2 px-3 sm:py-3 sm:px-4 block w-full border-gray-200 rounded-lg sm:text-sm focus:border-blue-500 focus:ring-blue-500 disabled:opacity-50 disabled:pointer-events-none dark:bg-neutral-900 dark:border-neutral-700 dark:text-neutral-400 dark:placeholder-neutral-500 dark:focus:ring-neutral-600" rows="3" data-hs-textarea-auto-height=""></textarea></div>"`;
